packages/taiko-client/prover/proof_producer/combined_producer.go

```diff
@@ -0,0 +1,126 @@
+package producer
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/taikoxyz/taiko-mono/packages/taiko-client/bindings/encoding"
+	"github.com/taikoxyz/taiko-mono/packages/taiko-client/bindings/metadata"
+)
+
+// CombinedProducer generates proofs from multiple producers in parallel and combines them.
+type CombinedProducer struct {
+	producers []ProofProducer
+	verifiers []common.Address
+	tier      uint16
+}
+
+// RequestProof implements the ProofProducer interface.
+func (c *CombinedProducer) RequestProof(
+	ctx context.Context,
+	opts *ProofRequestOptions,
+	blockID *big.Int,
+	meta metadata.TaikoBlockMetaData,
+	header *types.Header,
+	requestAt time.Time,
+) (*ProofWithHeader, error) {
+	var (
+		wg         sync.WaitGroup
+		proofChan  = make(chan []byte, len(c.producers))
+		errChan    = make(chan error, len(c.producers))
+		proofs     = make([][]byte, len(c.producers))
+		subProofs  = make([]encoding.SubProof, len(c.producers))
+		finalError error
+	)
+
+	for i, producer := range c.producers {
+		wg.Add(1)
+		go func(idx int, p ProofProducer) {
+			defer wg.Done()
+
+			proofWithHeader, err := p.RequestProof(ctx, opts, blockID, meta, header, requestAt)
+			if err != nil {
+				errChan <- fmt.Errorf("producer %d error: %w", idx, err)
+				return
+			}
+
+			proofChan <- proofWithHeader.Proof
+		}(i, producer)
+	}
+
+	wg.Wait()
+	close(proofChan)
+	close(errChan)
+
+	for err := range errChan {
+		if finalError == nil {
+			finalError = err
```

Comment:
Shouldn't we combine few errors into single one instead of taking the first one?